File: vscode-extension/src/resolutions.ts
import * as path from "node:path";

/** One flattened `<dependencyManagement>` entry: a concrete (already Scala-suffixed) artifact a BOM pins. */
export interface BomPin {
  organization: string;
  name: string;
  version: string;
}

/** A BOM's coordinate and its flattened pins, stored once and referenced by key from every project that sees it. */
export interface Bom {
  organization: string;
  name: string;
  version: string;
  entries: BomPin[];
}

/** A dependency whose `{{variable}}` version resolved, matched by organization + name + cross. */
export interface VariableResolution {
  organization: string;
  name: string;
  cross: boolean;
  variable: string;
  version: string;
}

/** One group's resolutions: the Scala binary versions it builds for, its visible BOMs (keys, precedence order) and its resolved variables. */
export interface ProjectResolutions {
  scalaBinaryVersions: string[];
  boms: string[];
  variables: VariableResolution[];
}

/** The parsed `.sbt-resolutions` dump written by the sbt plugin on load. */
export interface ResolutionsDump {
  version: number;
  sourceHash?: string;
  boms: Record<string, Bom>;
  projects: Record<string, ProjectResolutions>;
}

/** A resolved `*` version together with the BOM that pins it. */
export interface WildcardResolution {
  version: string;
  bom: { organization: string; name: string; version: string };
}

/** A resolved `{{variable}}` version together with the variable's name. */
export interface VariableLookupResult {
  version: string;
  variable: string;
}

/**
 * The read API consumed by the feature modules (decorations, hover, quickfixes), abstracting over the
 * {@link ResolutionsIndex} so tests can pass a fake. `stale` is `true` when the dump is older than the buffer it's
 * resolving against.
 */
export interface ResolutionLookup {
  resolveWildcard(group: string, org: string, name: string, isCross: boolean): WildcardResolution | undefined;
  resolveVariable(group: string, org: string, name: string, isCross: boolean): VariableLookupResult | undefined;
  pinFor(group: string, org: string, name: string, isCross: boolean): WildcardResolution | undefined;
  stale: boolean;
}

/** The reserved group name whose resolutions live in the meta-build dump. */
const SBT_BUILD_GROUP = "sbt-build";

/**
 * The two dump locations for a `dependencies.conf` at `<root>/project/dependencies.conf`: the main build's dump under
 * `<root>/target` and the meta-build's under `<root>/project/target`.
 */
export function dumpPathsFor(confFsPath: string): { main: string; meta: string } {
  const projectDir = path.dirname(confFsPath);
  const root = path.dirname(projectDir);
  const rel = path.join("target", "sbt-dependencies", ".sbt-resolutions");

  return { main: path.join(root, rel), meta: path.join(projectDir, rel) };
}

/** Parses a dump, returning `undefined` on malformed JSON or an unsupported `version` so callers silently disable. */
export function parseResolutionsDump(json: string): ResolutionsDump | undefined {
  try {
    const parsed = JSON.parse(json);
    if (parsed?.version !== 1 || typeof parsed.boms !== "object" || typeof parsed.projects !== "object") {
      return undefined;
    }
    return parsed as ResolutionsDump;
  } catch {
    return undefined;
  }
}

/**
 * Answers `*` and `{{variable}}` resolutions from the main and meta dumps, routing `sbt-build` to the meta dump and
 * every other group to the main one. Per-group `*` lookup maps are built on first use (a real BOM has 1000+ pins) with
 * first-BOM-wins precedence, matching the plugin.
 */
export class ResolutionsIndex {
  private readonly wildcardCache = new Map<string, Map<string, WildcardResolution>>();

  constructor(
    private readonly main: ResolutionsDump | undefined,
    private readonly meta: ResolutionsDump | undefined
  ) {}

  /** Whether either dump was loaded. */
  get hasData(): boolean {
    return this.main !== undefined || this.meta !== undefined;
  }

  private dumpFor(group: string): ResolutionsDump | undefined {
    return group === SBT_BUILD_GROUP ? this.meta : this.main;
  }

  private concreteName(name: string, isCross: boolean, sbv: string): string {
    return isCross ? `${name}_${sbv}` : name;
  }

  /** The first-wins `org:concreteArtifact` → resolution map for a group, built once and memoized. */
  private pinsFor(group: string): Map<string, WildcardResolution> {
    const cached = this.wildcardCache.get(group);
    if (cached) return cached;

    const map = new Map<string, WildcardResolution>();
    const dump = this.dumpFor(group);
    const project = dump?.projects[group];

    if (dump && project) {
      for (const key of project.boms) {
        const bom = dump.boms[key];
        if (!bom) continue;
        for (const pin of bom.entries) {
          const mapKey = `${pin.organization}:${pin.name}`;
          if (!map.has(mapKey)) {
            map.set(mapKey, {
              version: pin.version,
              bom: { organization: bom.organization, name: bom.name, version: bom.version },
            });
          }
        }
      }
    }

    this.wildcardCache.set(group, map);
    return map;
  }

  resolveWildcard(group: string, org: string, name: string, isCross: boolean): WildcardResolution | undefined {
    const dump = this.dumpFor(group);
    const project = dump?.projects[group];
    if (!project) return undefined;

    const sbv = project.scalaBinaryVersions[0];
    return this.pinsFor(group).get(`${org}:${this.concreteName(name, isCross, sbv)}`);
  }

  resolveVariable(group: string, org: string, name: string, isCross: boolean): VariableLookupResult | undefined {
    const dump = this.dumpFor(group);
    const project = dump?.projects[group];
    if (!project) return undefined;

    const match = project.variables.find(
      (v) => v.organization === org && v.name === name && v.cross === isCross
    );
    return match ? { version: match.version, variable: match.variable } : undefined;
  }

  pinFor(group: string, org: string, name: string, isCross: boolean): WildcardResolution | undefined {
    return this.resolveWildcard(group, org, name, isCross);
  }

  /** Wraps this index as a {@link ResolutionLookup} carrying the given staleness. */
  asLookup(stale: boolean): ResolutionLookup {
    return {
      resolveWildcard: this.resolveWildcard.bind(this),
      resolveVariable: this.resolveVariable.bind(this),
      pinFor: this.pinFor.bind(this),
      stale,
    };
  }
}

File: ide-plugins/intellij/src/main/scala/com/alejandrohdezma/sbt/dependencies/intellij/resolutions/Resolutions.scala
/*
 * Copyright 2025-2026 Alejandro Hernández <https://github.com/alejandrohdezma>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alejandrohdezma.sbt.dependencies.intellij.resolutions

import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap

import scala.jdk.CollectionConverters._
import scala.util.Try

import com.typesafe.config.ConfigFactory
import com.typesafe.config.ConfigObject

/** The `.sbt-resolutions` dump the sbt plugin writes on every load: each project's visible BOM pins and resolved
  * `{{variable}}` versions, keyed the same way the VSCode extension reads them.
  */
object Resolutions {

  /** One flattened `dependencyManagement` entry: a concrete (already Scala-suffixed) artifact a BOM pins. */
  final case class BomPin(organization: String, name: String, version: String)

  /** A BOM's coordinate and its flattened pins. */
  final case class Bom(organization: String, name: String, version: String, entries: List[BomPin])

  /** A dependency whose `{{variable}}` version resolved, matched by organization + name + cross. */
  final case class VariableResolution(
      organization: String,
      name: String,
      cross: Boolean,
      variable: String,
      version: String
  )

  /** One group's resolutions: its Scala binary versions, visible BOM keys (precedence order) and resolved variables. */
  final case class ProjectResolutions(
      scalaBinaryVersions: List[String],
      boms: List[String],
      variables: List[VariableResolution]
  )

  /** A parsed dump file. */
  final case class Dump(sourceHash: Option[String], boms: Map[String, Bom], projects: Map[String, ProjectResolutions])

  /** A resolved `*` version together with the BOM that pins it. */
  final case class Pin(version: String, bomOrganization: String, bomName: String, bomVersion: String)

  /** Parses a dump, returning `None` on malformed JSON or an unsupported `version` so callers silently disable. */
  def parse(json: String): Option[Dump] =
    Try {
      val config = ConfigFactory.parseString(json)

      Option.when(config.getInt("version") == 1) {
        val boms = config.getObject("boms").asScala.toMap.map { case (key, value) =>
          val bom = value.asInstanceOf[ConfigObject].toConfig

          key -> Bom(
            bom.getString("organization"),
            bom.getString("name"),
            bom.getString("version"),
            bom.getConfigList("entries").asScala.toList.map { entry =>
              BomPin(entry.getString("organization"), entry.getString("name"), entry.getString("version"))
            }
          )
        }

        val projects = config.getObject("projects").asScala.toMap.map { case (key, value) =>
          val project = value.asInstanceOf[ConfigObject].toConfig

          key -> ProjectResolutions(
            project.getStringList("scalaBinaryVersions").asScala.toList,
            project.getStringList("boms").asScala.toList,
            project.getConfigList("variables").asScala.toList.map { variable =>
              VariableResolution(
                variable.getString("organization"), variable.getString("name"), variable.getBoolean("cross"),
                variable.getString("variable"), variable.getString("version")
              )
            }
          )
        }

        Dump(Try(config.getString("sourceHash")).toOption, boms, projects)
      }
    }.toOption.flatten

  /** The lookup for the `dependencies.conf` at `conf` against its current `text`, or `None` when neither the main nor
    * the meta build has written a dump (feature off). Parsed dumps are cached and only re-read when their modification
    * time changes; staleness is an exact SHA-1 mismatch between `text` and the hash recorded in the dump.
    */
  def lookupFor(conf: Path, text: String): Option[Lookup] = {
    val relative = Path.of("target", "sbt-dependencies", ".sbt-resolutions")

    val main = cached(conf.getParent.getParent.resolve(relative))
    val meta = cached(conf.getParent.resolve(relative))

    Option.when(main.isDefined || meta.isDefined) {
      val sourceHash = main.flatMap(_.sourceHash).orElse(meta.flatMap(_.sourceHash))

      new Lookup(main, meta, stale = sourceHash.exists(_ != sha1(text)))
    }
  }

  /** Answers `*` and `{{variable}}` resolutions, routing the `sbt-build` group to the meta-build dump and every other
    * group to the main one, with first-BOM-wins precedence — matching the sbt plugin.
    */
  final class Lookup(main: Option[Dump], meta: Option[Dump], val stale: Boolean) {

    /** The BOM pin for a `*` version, if any BOM visible to `group` pins the artifact. */
    def resolveWildcard(group: String, organization: String, name: String, isCross: Boolean): Option[Pin] =
      dumpFor(group).flatMap { dump =>
        dump.projects.get(group).flatMap { project =>
          val concrete = project.scalaBinaryVersions.headOption
            .filter(_ => isCross)
            .fold(name)(sbv => s"${name}_$sbv")

          pinsFor(group, dump, project).get(s"$organization:$concrete")
        }
      }

    /** The resolved version and variable name for a `{{variable}}` dependency, if the dump recorded it. */
    def resolveVariable(
        group: String,
        organization: String,
        name: String,
        isCross: Boolean
    ): Option[VariableResolution] =
      dumpFor(group)
        .flatMap(_.projects.get(group))
        .flatMap(_.variables.find(v => v.organization == organization && v.name == name && v.cross == isCross))

    private def dumpFor(group: String): Option[Dump] = if (group == "sbt-build") meta else main

    private val pinsCache = new ConcurrentHashMap[String, Map[String, Pin]]()

    private def pinsFor(group: String, dump: Dump, project: ProjectResolutions): Map[String, Pin] =
      pinsCache.computeIfAbsent(
        group,
        _ =>
          project.boms
            .flatMap(key => dump.boms.get(key))
            .flatMap(bom =>
              bom.entries.map(pin =>
                s"${pin.organization}:${pin.name}" -> Pin(pin.version, bom.organization, bom.name, bom.version)
              )
            )
            .foldLeft(Map.empty[String, Pin]) { case (map, (key, pin)) =>
              if (map.contains(key)) map else map.updated(key, pin)
            }
      )

  }

  private val dumpCache = new ConcurrentHashMap[Path, (Long, Option[Dump])]()

  private def cached(path: Path): Option[Dump] = {
    val mtime = Try(Files.getLastModifiedTime(path).toMillis).getOrElse(0L)

    if (mtime == 0L) {
      dumpCache.remove(path)
      None
    } else
      dumpCache
        .compute(
          path,
          (_, previous) =>
            previous match {
              case (`mtime`, dump) => (mtime, dump)
              case _               => (mtime, Try(Files.readString(path, StandardCharsets.UTF_8)).toOption.flatMap(parse))
            }
        )
        ._2
  }

  private def sha1(text: String): String =
    MessageDigest
      .getInstance("SHA-1")
      .digest(text.getBytes(StandardCharsets.UTF_8))
      .map(byte => f"$byte%02x")
      .mkString

}
